File: apps/android/app/src/main/kotlin/dev/hyphen/android/transport/SessionHandshake.kt
package dev.hyphen.android.transport

import java.security.cert.X509Certificate
import javax.net.ssl.SSLSocket

/**
 * Hello exchange over an authenticated TLS socket (HYP-M2-013, protocol
 * v0 §4): the initiator sends `hello` (envelope sessionId null, payload
 * per capability.schema.json) and the responder replies `hello` carrying
 * the assigned sessionId in the envelope — the SAME id as before when a
 * valid resume token was presented (that equality IS the resume signal)
 * — plus a fresh single-use resume token for the next reconnect.
 *
 * v0 notes: handshake hellos travel with requiresAck=false (the reply is
 * the acknowledgment; doc-sync in M2-015), and seq restarts per
 * connection (hello=1, the session continues at 2).
 */
object SessionHandshake {

    const val HANDSHAKE_TIMEOUT_MS = 10_000
    const val CAPABILITY_NOTIFICATIONS = "notifications.v1"
    const val CAPABILITY_TRANSFER = "transfer.v1"
    const val CAPABILITY_TEXT = "text.v1"
    const val CAPABILITY_DIAGNOSTICS = "diagnostics.v1"

    private const val DEFAULT_TRANSFER_MAX_CHUNK_BYTES = 1_048_576
    private const val MIN_TRANSFER_MAX_CHUNK_BYTES = 1024
    private const val MAX_TRANSFER_MAX_CHUNK_BYTES = 2_097_152
    private val DEVICE_KINDS = setOf("android", "macos")
    private val APP_VERSION = Regex("^[0-9]+\\.[0-9]+\\.[0-9]+([\\-+][0-9A-Za-z.\\-]+)?$")
    private val CAPABILITY_NAME = Regex("^[a-z][a-z0-9]*\\.v[0-9]+$")

    data class DeviceInfo(
        val kind: String, // "android" | "macos"
        val appVersion: String,
        val deviceName: String? = null,
    )

    data class Result(
        val sessionId: String,
        /** Token to present on the NEXT reconnect (null if peer issued none). */
        val resumeToken: String?,
        val resumed: Boolean,
        val peerDevice: DeviceInfo?,
        val negotiatedCapabilities: NegotiatedCapabilities,
    )

    data class NegotiatedCapabilities(
        private val entries: Map<String, Json.Obj>,
    ) {
        fun contains(name: String): Boolean = entries.containsKey(name)

        fun notificationReplyEnabled(): Boolean =
            entries[CAPABILITY_NOTIFICATIONS]?.string("reply", "off")?.let { it != "off" } == true

        fun notificationDismissEnabled(): Boolean =
            entries[CAPABILITY_NOTIFICATIONS]?.bool("dismiss", false) == true

        fun notificationPrivacyPolicyEnabled(): Boolean =
            entries[CAPABILITY_NOTIFICATIONS]?.bool("privacyPolicy", false) == true

        fun transferMaxChunkBytes(): Int? =
            ((entries[CAPABILITY_TRANSFER]?.get("maxChunkBytes") as? Json.Num)?.asLong())?.toInt()

        /** The negotiated text.v1 direction (null when text was not negotiated). */
        fun textDirection(): String? =
            (entries[CAPABILITY_TEXT]?.get("direction") as? Json.Str)?.value

        /**
         * Finalize text.v1 direction once device kinds are known (responder
         * side). Direction is expressed from the canonical android-endpoint
         * frame; the client adopts this verbatim. Drops the text capability
         * entirely when no compatible flow exists (review dim 05-02).
         */
        fun resolveTextDirection(
            localKind: String,
            localDir: String,
            peerKind: String?,
            peerDir: String?,
        ): NegotiatedCapabilities {
            if (!contains(CAPABILITY_TEXT)) return this
            val direction = negotiateTextDirection(
                localKind,
                localDir,
                peerKind ?: localKind,
                peerDir ?: "bidirectional",
            )
            val updated = LinkedHashMap(entries)
            if (direction == "none") {
                updated.remove(CAPABILITY_TEXT)
            } else {
                updated[CAPABILITY_TEXT] = Json.obj("direction" to Json.Str(direction))
            }
            return NegotiatedCapabilities(updated)
        }

        fun toJson(): Json.Obj = Json.Obj(entries)

        fun intersect(peer: NegotiatedCapabilities): NegotiatedCapabilities {
            val result = linkedMapOf<String, Json.Obj>()
            if (contains(CAPABILITY_NOTIFICATIONS) && peer.contains(CAPABILITY_NOTIFICATIONS)) {
                val local = entries.getValue(CAPABILITY_NOTIFICATIONS)
                val remote = peer.entries.getValue(CAPABILITY_NOTIFICATIONS)
                result[CAPABILITY_NOTIFICATIONS] = Json.obj(
                    "reply" to Json.Str(minReply(local.string("reply", "off"), remote.string("reply", "off"))),
                    "dismiss" to Json.Bool(local.bool("dismiss", false) && remote.bool("dismiss", false)),
                    "privacyPolicy" to Json.Bool(local.bool("privacyPolicy", false) && remote.bool("privacyPolicy", false)),
                )
            }
            if (contains(CAPABILITY_TRANSFER) && peer.contains(CAPABILITY_TRANSFER)) {
                val local = entries.getValue(CAPABILITY_TRANSFER)
                val remote = peer.entries.getValue(CAPABILITY_TRANSFER)
                result[CAPABILITY_TRANSFER] = Json.obj(
                    "resume" to Json.Bool(local.bool("resume", false) && remote.bool("resume", false)),
                    "maxChunkBytes" to Json.Num(
                        minOf(
                            local.int("maxChunkBytes", DEFAULT_TRANSFER_MAX_CHUNK_BYTES),
                            remote.int("maxChunkBytes", DEFAULT_TRANSFER_MAX_CHUNK_BYTES),
                        ).toString(),
                    ),
                )
            }
            if (contains(CAPABILITY_TEXT) && peer.contains(CAPABILITY_TEXT)) {
                // Carry the left operand's direction through instead of
                // fabricating bidirectional (review dim 05-02). The responder
                // finalizes the real direction via resolveTextDirection() once
                // device kinds are known; the client adopts the responder's
                // already-resolved value (its hello reply is the left operand).
                result[CAPABILITY_TEXT] = Json.obj(
                    "direction" to Json.Str(entries.getValue(CAPABILITY_TEXT).string("direction", "bidirectional")),
                )
            }
            if (contains(CAPABILITY_DIAGNOSTICS) && peer.contains(CAPABILITY_DIAGNOSTICS)) {
                val local = entries.getValue(CAPABILITY_DIAGNOSTICS)
                val remote = peer.entries.getValue(CAPABILITY_DIAGNOSTICS)
                result[CAPABILITY_DIAGNOSTICS] = Json.obj(
                    "redactedExport" to Json.Bool(local.bool("redactedExport", false) && remote.bool("redactedExport", false)),
                )
            }
            return NegotiatedCapabilities(result)
        }

        private fun Json.Obj.bool(key: String, default: Boolean): Boolean =
            (this[key] as? Json.Bool)?.value ?: default

        private fun Json.Obj.string(key: String, default: String): String =
            (this[key] as? Json.Str)?.value ?: default

        private fun Json.Obj.int(key: String, default: Int): Int =
            ((this[key] as? Json.Num)?.asLong())?.toInt() ?: default

        private fun minReply(left: String, right: String): String {
            val rank = mapOf("off" to 0, "beta" to 1, "on" to 2)
            val minRank = minOf(rank.getValue(left), rank.getValue(right))
            return rank.entries.first { it.value == minRank }.key
        }

        companion object {
            fun advertised(maxTransferChunkBytes: Int = DEFAULT_TRANSFER_MAX_CHUNK_BYTES): NegotiatedCapabilities {
                require(maxTransferChunkBytes in MIN_TRANSFER_MAX_CHUNK_BYTES..MAX_TRANSFER_MAX_CHUNK_BYTES) {
                    "transfer maxChunkBytes out of range"
                }
                return NegotiatedCapabilities(
                    linkedMapOf(
                        CAPABILITY_NOTIFICATIONS to Json.obj(
                            "reply" to Json.Str("beta"),
                            "dismiss" to Json.Bool(true),
                            "privacyPolicy" to Json.Bool(true),
                        ),
                        CAPABILITY_TRANSFER to Json.obj(
                            "resume" to Json.Bool(true),
                            "maxChunkBytes" to Json.Num(maxTransferChunkBytes.toString()),
                        ),
                        CAPABILITY_TEXT to Json.obj("direction" to Json.Str("bidirectional")),
                        CAPABILITY_DIAGNOSTICS to Json.obj("redactedExport" to Json.Bool(true)),
                    ),
                )
            }

            fun empty(): NegotiatedCapabilities = NegotiatedCapabilities(emptyMap())

            /**
             * Negotiate text.v1 direction from the canonical android-endpoint
             * frame. Each advertised `direction` is the advertiser's own
             * perspective (what it will do); v0 assumes exactly one android +
             * one macos peer, so the android device is the frame subject
             * (fallback: the first argument when neither is android). Returns
             * `bidirectional` / `send-only` / `receive-only` / `none`.
             */
            fun negotiateTextDirection(kindA: String, dirA: String, kindB: String, dirB: String): String {
                val aIsSubject = !(kindB == "android" && kindA != "android")
                val subjectDir = if (aIsSubject) dirA else dirB
                val otherDir = if (aIsSubject) dirB else dirA
                fun sends(d: String) = d == "bidirectional" || d == "send-only"
                fun receives(d: String) = d == "bidirectional" || d == "receive-only"
                val subjectSends = sends(subjectDir) && receives(otherDir)
                val subjectReceives = sends(otherDir) && receives(subjectDir)
                return when {
                    subjectSends && subjectReceives -> "bidirectional"
                    subjectSends -> "send-only"
                    subjectReceives -> "receive-only"
                    else -> "none"
                }
            }
        }
    }

    class HandshakeException(val code: String, message: String) : Exception(message)

    /** Client side: returns once the responder's hello arrives. */
    fun initiate(
        socket: SSLSocket,
        device: DeviceInfo,
        resumeToken: String?,
        previousSessionId: String?,
    ): Result {
        socket.soTimeout = HANDSHAKE_TIMEOUT_MS
        val hello = Envelope(
            messageId = Ulid.generate(),
            sessionId = null,
            type = Envelope.TYPE_HELLO,
            seq = 1,
            sentAtUnixMs = System.currentTimeMillis(),
            requiresAck = false,
            payload = helloPayload(device, resumeToken),
        )
        FrameIO.write(socket.outputStream, hello.encode())

        val reply = readHello(socket)
        socket.soTimeout = 0 // the session layer owns read timeouts from here
        val sessionId = reply.sessionId
            ?: throw HandshakeException("protocol/invalid-envelope", "responder hello carried no sessionId")
        return Result(
            sessionId = sessionId,
            resumeToken = (reply.payload["resumeToken"] as? Json.Str)?.value,
            resumed = sessionId == previousSessionId,
            peerDevice = deviceOf(reply.payload),
            negotiatedCapabilities = capabilitiesOf(reply.payload).intersect(NegotiatedCapabilities.advertised()),
        )
    }

    /** Server side: consumes the initiator's hello, assigns or resumes a
     *  session, replies, and returns what the session layer needs. */
    fun respond(
        socket: SSLSocket,
        device: DeviceInfo,
        tokenStore: ResumeTokenStore,
    ): Result {
        socket.soTimeout = HANDSHAKE_TIMEOUT_MS
        val peerFingerprint = TlsIdentity.spkiFingerprintOf(
            socket.session.peerCertificates.first() as X509Certificate
        )

        val hello = readHello(socket)
        val peerCaps = capabilitiesOf(hello.payload)
        val peerDevice = deviceOf(hello.payload)
        val advertised = NegotiatedCapabilities.advertised()
        val negotiatedCapabilities = advertised.intersect(peerCaps).resolveTextDirection(
            localKind = device.kind,
            localDir = advertised.textDirection() ?: "bidirectional",
            peerKind = peerDevice?.kind,
            peerDir = peerCaps.textDirection(),
        )
        val presented = (hello.payload["resumeToken"] as? Json.Str)?.value
        val resumedSessionId = presented?.let { tokenStore.redeem(it, peerFingerprint) }
        val sessionId = resumedSessionId ?: "s_${Ulid.generate()}"
        val nextToken = tokenStore.issue(sessionId, peerFingerprint)

        val reply = Envelope(
            messageId = Ulid.generate(),
            sessionId = sessionId,
            type = Envelope.TYPE_HELLO,
            seq = 1,
            sentAtUnixMs = System.currentTimeMillis(),
            requiresAck = false,
            payload = helloPayload(device, nextToken, negotiatedCapabilities),
        )
        FrameIO.write(socket.outputStream, reply.encode())
        socket.soTimeout = 0 // the session layer owns read timeouts from here

        return Result(
            sessionId = sessionId,
            resumeToken = nextToken,
            resumed = resumedSessionId != null,
            peerDevice = deviceOf(hello.payload),
            negotiatedCapabilities = negotiatedCapabilities,
        )
    }

    private fun readHello(socket: SSLSocket): Envelope {
        val frame = FrameIO.read(socket.inputStream)
            ?: throw HandshakeException("transport/connection-lost", "peer closed before hello")
        val envelope = try {
            Envelope.decode(frame)
        } catch (e: EnvelopeException) {
            throw HandshakeException("protocol/invalid-envelope", e.message ?: "bad hello")
        }
        if (envelope.type != Envelope.TYPE_HELLO) {
            throw HandshakeException("protocol/invalid-envelope", "expected hello, got ${envelope.type}")
        }
        validateHelloPayload(envelope.payload)
        return envelope
    }

    /** Structural check per capability.schema.json (strict top level). */
    private fun validateHelloPayload(payload: Json.Obj) {
        val unknown = payload.entries.keys - setOf("device", "resumeToken", "capabilities")
        if (unknown.isNotEmpty()) {
            throw HandshakeException("protocol/invalid-envelope", "unknown hello fields: $unknown")
        }
        val device = payload["device"] as? Json.Obj
            ?: throw HandshakeException("protocol/invalid-envelope", "hello.device missing")
        val unknownDevice = device.entries.keys - setOf("kind", "appVersion", "osVersion", "deviceName")
        if (unknownDevice.isNotEmpty()) {
            throw HandshakeException("protocol/invalid-envelope", "unknown hello.device fields: $unknownDevice")
        }
        if (device["kind"] !is Json.Str || device["appVersion"] !is Json.Str) {
            throw HandshakeException("protocol/invalid-envelope", "hello.device incomplete")
        }
        val kind = (device["kind"] as Json.Str).value
        if (kind !in DEVICE_KINDS) {
            throw HandshakeException("protocol/invalid-envelope", "hello.device.kind invalid")
        }
        val appVersion = (device["appVersion"] as Json.Str).value
        if (!APP_VERSION.matches(appVersion)) {
            throw HandshakeException("protocol/invalid-envelope", "hello.device.appVersion invalid")
        }
        when (payload["resumeToken"]) {
            is Json.Null, is Json.Str -> Unit
            else -> throw HandshakeException("protocol/invalid-envelope", "hello.resumeToken must be string or null")
        }
        if (payload["capabilities"] !is Json.Obj) {
            throw HandshakeException("protocol/invalid-envelope", "hello.capabilities missing")
        }
        capabilitiesOf(payload)
    }

    private fun helloPayload(
        device: DeviceInfo,
        resumeToken: String?,
        capabilities: NegotiatedCapabilities = NegotiatedCapabilities.advertised(),
    ): Json.Obj {
        val deviceEntries = linkedMapOf<String, Json>(
            "kind" to Json.Str(device.kind),
            "appVersion" to Json.Str(device.appVersion),
        )
        device.deviceName?.let { deviceEntries["deviceName"] = Json.Str(it) }
        return Json.obj(
            "device" to Json.Obj(deviceEntries),
            "resumeToken" to (resumeToken?.let { Json.Str(it) } ?: Json.Null),
            "capabilities" to capabilities.toJson(),
        )
    }

    private fun capabilitiesOf(payload: Json.Obj): NegotiatedCapabilities {
        val raw = payload["capabilities"] as? Json.Obj
            ?: throw HandshakeException("protocol/invalid-envelope", "hello.capabilities missing")
        val result = linkedMapOf<String, Json.Obj>()
        for ((name, value) in raw.entries) {
            if (!CAPABILITY_NAME.matches(name)) {
                throw HandshakeException("protocol/invalid-envelope", "bad capability '$name'")
            }
            val options = value as? Json.Obj
                ?: throw HandshakeException("protocol/invalid-envelope", "capability '$name' options must be object")
            validateCapabilityOptions(name, options)
            result[name] = options
        }
        return NegotiatedCapabilities(result)
    }

    private fun validateCapabilityOptions(name: String, options: Json.Obj) {
        when (name) {
            CAPABILITY_NOTIFICATIONS -> {
                options["reply"]?.let { raw ->
                    val reply = (raw as? Json.Str)?.value
                        ?: throw HandshakeException("protocol/invalid-envelope", "notifications.v1.reply must be string")
                    if (reply !in setOf("off", "beta", "on")) {
                        throw HandshakeException("protocol/invalid-envelope", "notifications.v1.reply invalid")
                    }
                }
                if (options["dismiss"] != null && options["dismiss"] !is Json.Bool) {
                    throw HandshakeException("protocol/invalid-envelope", "notifications.v1.dismiss must be boolean")
                }
                if (options["privacyPolicy"] != null && options["privacyPolicy"] !is Json.Bool) {
                    throw HandshakeException("protocol/invalid-envelope", "notifications.v1.privacyPolicy must be boolean")
                }
            }
            CAPABILITY_TRANSFER -> {
                if (options["resume"] != null && options["resume"] !is Json.Bool) {
                    throw HandshakeException("protocol/invalid-envelope", "transfer.v1.resume must be boolean")
                }
                options["maxChunkBytes"]?.let { raw ->
                    val max = (raw as? Json.Num)?.asLong()
                        ?: throw HandshakeException("protocol/invalid-envelope", "transfer.v1.maxChunkBytes must be integer")
                    if (max < MIN_TRANSFER_MAX_CHUNK_BYTES.toLong() || max > MAX_TRANSFER_MAX_CHUNK_BYTES.toLong()) {
                        throw HandshakeException("protocol/invalid-envelope", "transfer.v1.maxChunkBytes out of range")
                    }
                }
            }
            CAPABILITY_TEXT -> {
                options["direction"]?.let { raw ->
                    val direction = (raw as? Json.Str)?.value
                        ?: throw HandshakeException("protocol/invalid-envelope", "text.v1.direction must be string")
                    if (direction !in setOf("bidirectional", "send-only", "receive-only")) {
                        throw HandshakeException("protocol/invalid-envelope", "text.v1.direction invalid")
                    }
                }
            }
            CAPABILITY_DIAGNOSTICS -> {
                if (options["redactedExport"] != null && options["redactedExport"] !is Json.Bool) {
                    throw HandshakeException("protocol/invalid-envelope", "diagnostics.v1.redactedExport must be boolean")
                }
            }
        }
    }

    private fun deviceOf(payload: Json.Obj): DeviceInfo? {
        val device = payload["device"] as? Json.Obj ?: return null
        return DeviceInfo(
            kind = (device["kind"] as? Json.Str)?.value ?: return null,
            appVersion = (device["appVersion"] as? Json.Str)?.value ?: return null,
            deviceName = (device["deviceName"] as? Json.Str)?.value,
        )
    }
}

File: apps/macos/Sources/HyphenTransport/SessionHandshake.swift
import Foundation
import Network

/// Hello exchange over an authenticated NWConnection (HYP-M2-013,
/// protocol v0 §4): the initiator sends `hello` (envelope sessionId
/// null) and the responder replies `hello` carrying the assigned
/// sessionId in the envelope — the SAME id as before when a valid resume
/// token was presented (that equality IS the resume signal) — plus a
/// fresh single-use resume token for the next reconnect. Twin of the
/// Android `SessionHandshake`.
///
/// v0 notes: handshake hellos travel with requiresAck=false (the reply
/// is the acknowledgment; doc-sync in M2-015) and seq restarts per
/// connection (hello=1, the session continues at 2).
public enum SessionHandshake {

    public static let timeoutMs = 10_000
    public static let capabilityNotifications = "notifications.v1"
    public static let capabilityTransfer = "transfer.v1"
    public static let capabilityText = "text.v1"
    public static let capabilityDiagnostics = "diagnostics.v1"

    private static let defaultTransferMaxChunkBytes = 1_048_576
    private static let minTransferMaxChunkBytes = 1_024
    private static let maxTransferMaxChunkBytes = 2_097_152
    private static let deviceKinds: Set<String> = ["android", "macos"]
    private static let appVersionPattern = "^[0-9]+\\.[0-9]+\\.[0-9]+([\\-+][0-9A-Za-z.\\-]+)?$"
    private static let capabilityPattern = "^[a-z][a-z0-9]*\\.v[0-9]+$"

    public struct DeviceInfo {
        public let kind: String // "android" | "macos"
        public let appVersion: String
        public let deviceName: String?

        public init(kind: String, appVersion: String, deviceName: String? = nil) {
            self.kind = kind
            self.appVersion = appVersion
            self.deviceName = deviceName
        }
    }

    public struct HandshakeResult {
        public let sessionId: String
        /// Token to present on the NEXT reconnect (nil if none issued).
        public let resumeToken: String?
        public let resumed: Bool
        public let peerDeviceName: String?
        public let negotiatedCapabilities: NegotiatedCapabilities
        /// Bytes the handshake read past the hello — the session must
        /// replay them before reading from the connection.
        public let leftover: Data
    }

    public struct NegotiatedCapabilities {
        private let entries: [String: [String: Any]]

        public init(_ entries: [String: [String: Any]]) {
            self.entries = entries
        }

        public func contains(_ capability: String) -> Bool {
            entries[capability] != nil
        }

        public var transferMaxChunkBytes: Int? {
            entries[capabilityTransfer]?["maxChunkBytes"] as? Int
        }

        public var wireObject: [String: Any] {
            entries
        }

        public func intersecting(_ peer: NegotiatedCapabilities) -> NegotiatedCapabilities {
            var result: [String: [String: Any]] = [:]
            if contains(capabilityNotifications), peer.contains(capabilityNotifications),
               let local = entries[capabilityNotifications],
               let remote = peer.entries[capabilityNotifications] {
                result[capabilityNotifications] = [
                    "reply": Self.minReply(
                        local["reply"] as? String ?? "off",
                        remote["reply"] as? String ?? "off"
                    ),
                    "dismiss": (local["dismiss"] as? Bool ?? false) && (remote["dismiss"] as? Bool ?? false),
                ]
            }
            if contains(capabilityTransfer), peer.contains(capabilityTransfer),
               let local = entries[capabilityTransfer],
               let remote = peer.entries[capabilityTransfer] {
                result[capabilityTransfer] = [
                    "resume": (local["resume"] as? Bool ?? false) && (remote["resume"] as? Bool ?? false),
                    "maxChunkBytes": min(
                        local["maxChunkBytes"] as? Int ?? defaultTransferMaxChunkBytes,
                        remote["maxChunkBytes"] as? Int ?? defaultTransferMaxChunkBytes
                    ),
                ]
            }
            if contains(capabilityText), peer.contains(capabilityText) {
                result[capabilityText] = ["direction": "bidirectional"]
            }
            if contains(capabilityDiagnostics), peer.contains(capabilityDiagnostics),
               let local = entries[capabilityDiagnostics],
               let remote = peer.entries[capabilityDiagnostics] {
                result[capabilityDiagnostics] = [
                    "redactedExport": (local["redactedExport"] as? Bool ?? false) && (remote["redactedExport"] as? Bool ?? false),
                ]
            }
            return NegotiatedCapabilities(result)
        }

        public static func advertised(maxTransferChunkBytes: Int = 1_048_576) throws -> NegotiatedCapabilities {
            guard (minTransferMaxChunkBytes...maxTransferMaxChunkBytes).contains(maxTransferChunkBytes) else {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "transfer maxChunkBytes out of range")
            }
            return NegotiatedCapabilities([
                capabilityNotifications: ["reply": "beta", "dismiss": true],
                capabilityTransfer: ["resume": true, "maxChunkBytes": maxTransferChunkBytes],
                capabilityText: ["direction": "bidirectional"],
                capabilityDiagnostics: ["redactedExport": true],
            ])
        }

        public static func empty() -> NegotiatedCapabilities {
            NegotiatedCapabilities([:])
        }

        private static func minReply(_ left: String, _ right: String) -> String {
            let rank = ["off": 0, "beta": 1, "on": 2]
            let minRank = min(rank[left] ?? 0, rank[right] ?? 0)
            return rank.first { $0.value == minRank }?.key ?? "off"
        }
    }

    public struct HandshakeError: Error, CustomStringConvertible {
        public let code: String
        public let detail: String
        public var description: String { "\(code): \(detail)" }
    }

    /// Client side. Completion fires exactly once on `queue`.
    public static func initiate(
        connection: NWConnection,
        device: DeviceInfo,
        resumeToken: String?,
        previousSessionId: String?,
        queue: DispatchQueue,
        completion: @escaping (Result<HandshakeResult, Error>) -> Void
    ) {
        let hello = Envelope(
            messageId: Ulid.generate(),
            sessionId: nil,
            type: Envelope.typeHello,
            seq: 1,
            sentAtUnixMs: Int64(Date().timeIntervalSince1970 * 1000),
            requiresAck: false,
            payload: helloPayload(device: device, resumeToken: resumeToken)
        )
        sendEnvelope(hello, over: connection) { sendError in
            if let sendError {
                completion(.failure(sendError))
                return
            }
            readHello(connection: connection, queue: queue) { result in
                completion(result.flatMap { reply, leftover in
                    guard let sessionId = reply.sessionId else {
                        return .failure(HandshakeError(
                            code: "protocol/invalid-envelope",
                            detail: "responder hello carried no sessionId"
                        ))
                    }
                    let negotiatedCapabilities: NegotiatedCapabilities
                    do {
                        negotiatedCapabilities = try capabilities(of: reply.payload)
                            .intersecting(try NegotiatedCapabilities.advertised())
                    } catch {
                        return .failure(error)
                    }
                    return .success(HandshakeResult(
                        sessionId: sessionId,
                        resumeToken: (reply.payload["resumeToken"] as? String),
                        resumed: sessionId == previousSessionId,
                        peerDeviceName: deviceName(of: reply.payload),
                        negotiatedCapabilities: negotiatedCapabilities,
                        leftover: leftover
                    ))
                })
            }
        }
    }

    /// Server side: consumes the initiator's hello, assigns or resumes a
    /// session, replies. `peerFingerprint` comes from the listener's
    /// verify callback (it authenticated this connection).
    public static func respond(
        connection: NWConnection,
        device: DeviceInfo,
        peerFingerprint: Data,
        tokenStore: ResumeTokenStore,
        queue: DispatchQueue,
        completion: @escaping (Result<HandshakeResult, Error>) -> Void
    ) {
        readHello(connection: connection, queue: queue) { result in
            switch result {
            case .failure(let error):
                completion(.failure(error))
            case .success(let (hello, leftover)):
                let negotiatedCapabilities: NegotiatedCapabilities
                do {
                    negotiatedCapabilities = try NegotiatedCapabilities.advertised()
                        .intersecting(try capabilities(of: hello.payload))
                } catch {
                    completion(.failure(error))
                    return
                }
                let presented = hello.payload["resumeToken"] as? String
                let resumedSessionId = presented.flatMap {
                    tokenStore.redeem(token: $0, peerFingerprint: peerFingerprint)
                }
                let sessionId = resumedSessionId ?? "s_\(Ulid.generate())"
                let nextToken = tokenStore.issue(sessionId: sessionId, peerFingerprint: peerFingerprint)

                let reply = Envelope(
                    messageId: Ulid.generate(),
                    sessionId: sessionId,
                    type: Envelope.typeHello,
                    seq: 1,
                    sentAtUnixMs: Int64(Date().timeIntervalSince1970 * 1000),
                    requiresAck: false,
                    payload: helloPayload(device: device, resumeToken: nextToken, capabilities: negotiatedCapabilities)
                )
                sendEnvelope(reply, over: connection) { sendError in
                    if let sendError {
                        completion(.failure(sendError))
                        return
                    }
                    completion(.success(HandshakeResult(
                        sessionId: sessionId,
                        resumeToken: nextToken,
                        resumed: resumedSessionId != nil,
                        peerDeviceName: deviceName(of: hello.payload),
                        negotiatedCapabilities: negotiatedCapabilities,
                        leftover: leftover
                    )))
                }
            }
        }
    }

    // MARK: - Internals

    private static func sendEnvelope(
        _ envelope: Envelope,
        over connection: NWConnection,
        completion: @escaping (Error?) -> Void
    ) {
        do {
            let frame = try FrameCodec.encode(try envelope.encode())
            connection.send(content: frame, completion: .contentProcessed { completion($0) })
        } catch {
            completion(error)
        }
    }

    private static func readHello(
        connection: NWConnection,
        queue: DispatchQueue,
        completion: @escaping (Result<(Envelope, Data), Error>) -> Void
    ) {
        let reader = FrameReader()
        var done = false
        let finish: (Result<(Envelope, Data), Error>) -> Void = { result in
            guard !done else { return }
            done = true
            completion(result)
        }
        queue.asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) {
            finish(.failure(HandshakeError(code: "transport/connection-lost", detail: "hello timed out")))
        }

        func receiveNext() {
            connection.receive(minimumIncompleteLength: 1, maximumLength: 65_536) { data, _, isComplete, error in
                guard !done else { return }
                if let data, !data.isEmpty {
                    do {
                        if let first = try reader.feedUntilFirst(data) {
                            let envelope = try Envelope.decode(first.frame)
                            guard envelope.type == Envelope.typeHello else {
                                throw HandshakeError(
                                    code: "protocol/invalid-envelope",
                                    detail: "expected hello, got \(envelope.type)"
                                )
                            }
                            try validateHelloPayload(envelope.payload)
                            finish(.success((envelope, first.leftover)))
                            return
                        }
                    } catch {
                        finish(.failure(error))
                        return
                    }
                }
                if isComplete || error != nil {
                    finish(.failure(HandshakeError(
                        code: "transport/connection-lost",
                        detail: "peer closed before hello"
                    )))
                    return
                }
                receiveNext()
            }
        }
        receiveNext()
    }

    /// Structural check per capability.schema.json (strict top level).
    private static func validateHelloPayload(_ payload: [String: Any]) throws {
        let unknown = Set(payload.keys).subtracting(["device", "resumeToken", "capabilities"])
        guard unknown.isEmpty else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "unknown hello fields: \(unknown.sorted())")
        }
        guard let device = payload["device"] as? [String: Any],
              device["kind"] is String, device["appVersion"] is String
        else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.device missing or incomplete")
        }
        let unknownDevice = Set(device.keys).subtracting(["kind", "appVersion", "osVersion", "deviceName"])
        guard unknownDevice.isEmpty else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "unknown hello.device fields: \(unknownDevice.sorted())")
        }
        guard let kind = device["kind"] as? String, deviceKinds.contains(kind) else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.device.kind invalid")
        }
        guard let appVersion = device["appVersion"] as? String,
              appVersion.range(of: appVersionPattern, options: .regularExpression) != nil
        else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.device.appVersion invalid")
        }
        if let token = payload["resumeToken"], !(token is NSNull), !(token is String) {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.resumeToken must be string or null")
        }
        guard payload["resumeToken"] != nil else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.resumeToken missing")
        }
        guard payload["capabilities"] is [String: Any] else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.capabilities missing")
        }
        _ = try capabilities(of: payload)
    }

    private static func helloPayload(
        device: DeviceInfo,
        resumeToken: String?,
        capabilities: NegotiatedCapabilities = try! NegotiatedCapabilities.advertised()
    ) -> [String: Any] {
        var deviceDict: [String: Any] = [
            "kind": device.kind,
            "appVersion": device.appVersion,
        ]
        if let deviceName = device.deviceName { deviceDict["deviceName"] = deviceName }
        return [
            "device": deviceDict,
            "resumeToken": resumeToken ?? NSNull(),
            "capabilities": capabilities.wireObject,
        ]
    }

    private static func capabilities(of payload: [String: Any]) throws -> NegotiatedCapabilities {
        guard let raw = payload["capabilities"] as? [String: Any] else {
            throw HandshakeError(code: "protocol/invalid-envelope", detail: "hello.capabilities missing")
        }
        var result: [String: [String: Any]] = [:]
        for (name, value) in raw {
            guard name.range(of: capabilityPattern, options: .regularExpression) != nil else {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "bad capability '\(name)'")
            }
            guard let options = value as? [String: Any] else {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "capability '\(name)' options must be object")
            }
            try validateCapabilityOptions(name: name, options: options)
            result[name] = options
        }
        return NegotiatedCapabilities(result)
    }

    private static func validateCapabilityOptions(name: String, options: [String: Any]) throws {
        switch name {
        case capabilityNotifications:
            if let raw = options["reply"] {
                guard let reply = raw as? String else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "notifications.v1.reply must be string")
                }
                guard ["off", "beta", "on"].contains(reply) else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "notifications.v1.reply invalid")
                }
            }
            if let raw = options["dismiss"], !isBoolean(raw) {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "notifications.v1.dismiss must be boolean")
            }
        case capabilityTransfer:
            if let raw = options["resume"], !isBoolean(raw) {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "transfer.v1.resume must be boolean")
            }
            if let raw = options["maxChunkBytes"] {
                guard let max = raw as? Int else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "transfer.v1.maxChunkBytes must be integer")
                }
                guard (minTransferMaxChunkBytes...maxTransferMaxChunkBytes).contains(max) else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "transfer.v1.maxChunkBytes out of range")
                }
            }
        case capabilityText:
            if let raw = options["direction"] {
                guard let direction = raw as? String else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "text.v1.direction must be string")
                }
                guard ["bidirectional", "send-only", "receive-only"].contains(direction) else {
                    throw HandshakeError(code: "protocol/invalid-envelope", detail: "text.v1.direction invalid")
                }
            }
        case capabilityDiagnostics:
            if let raw = options["redactedExport"], !isBoolean(raw) {
                throw HandshakeError(code: "protocol/invalid-envelope", detail: "diagnostics.v1.redactedExport must be boolean")
            }
        default:
            break
        }
    }

    private static func deviceName(of payload: [String: Any]) -> String? {
        (payload["device"] as? [String: Any])?["deviceName"] as? String
    }

    private static func isBoolean(_ value: Any) -> Bool {
        CFGetTypeID(value as CFTypeRef) == CFBooleanGetTypeID()
    }
}
